core/blockchain/production/impl/block_producer_impl.cpp

```diff
@@ -0,0 +1,151 @@
+  using storage::amt::Root;
+  using storage::ipfs::InMemoryDatastore;
+  using vm::message::SignedMessage;
+  using vm::message::UnsignedMessage;
+
+  BlockProducerImpl::BlockProducerImpl(
+      std::shared_ptr<IpfsDatastore> data_store,
+      std::shared_ptr<MessageStorage> message_store,
+      std::shared_ptr<UTCClock> utc_clock,
+      std::shared_ptr<ChainEpochClock> epoch_clock,
+      std::shared_ptr<WeightCalculator> weight_calculator,
+      std::shared_ptr<BlsProvider> crypto_provider,
+      std::shared_ptr<Interpreter> interpreter)
+      : data_storage_{std::move(data_store)},
+        message_storage_{std::move(message_store)},
+        clock_{std::move(utc_clock)},
+        epoch_{std::move(epoch_clock)},
+        chain_weight_calculator_{std::move(weight_calculator)},
+        bls_provider_{std::move(crypto_provider)},
+        vm_interpreter_{std::move(interpreter)} {}
+
+  outcome::result<BlockProducer::Block> BlockProducerImpl::generate(
+      primitives::address::Address miner_address,
+      const CID &parent_tipset_id,
+      EPostProof proof,
+      Ticket ticket,
+      std::shared_ptr<Indices> indices) {
+    OUTCOME_TRY(parent_tipset, getTipset(parent_tipset_id));
+    OUTCOME_TRY(
+        vm_result,
+        vm_interpreter_->interpret(data_storage_, parent_tipset, indices));
+    OUTCOME_TRY(parent_weight,
+                chain_weight_calculator_->calculateWeight(parent_tipset));
+    std::vector<SignedMessage> messages =
+        message_storage_->getTopScored(config::kBlockMaxMessagesCount);
+    OUTCOME_TRY(msg_meta, getMessagesMeta(messages));
+    std::vector<UnsignedMessage> bls_messages;
+    std::vector<SignedMessage> secp_messages;
+    std::vector<crypto::bls::Signature> bls_signatures;
+    for (auto &&message : messages) {
+      visit_in_place(
+          message.signature,
+          [&message, &bls_messages, &bls_signatures](
+              const BlsSignature &signature) {
+            bls_messages.emplace_back(std::move(message.message));
+            bls_signatures.push_back(signature);
+          },
+          [&message, &secp_messages](const Secp256k1Signature &signature) {
+            secp_messages.emplace_back(std::move(message));
+          });
+    }
+    OUTCOME_TRY(bls_aggregate_sign,
+                bls_provider_->aggregateSignatures(bls_signatures));
+    Time now = clock_->nowUTC();
+    OUTCOME_TRY(current_epoch, epoch_->epochAtTime(now));
+    BlockHeader header{
+        .miner = std::move(miner_address),
+        .ticket = ticket,
+        .epost_proof = std::move(proof),
+        .parents = std::move(parent_tipset.cids),
+        .parent_weight = parent_weight,
+        .height = static_cast<uint64_t>(current_epoch),
+        .parent_state_root = std::move(vm_result.state_root),
+        .parent_message_receipts = std::move(vm_result.message_receipts),
+        .messages = msg_meta.getCID(),
+        .bls_aggregate = {bls_aggregate_sign.begin(), bls_aggregate_sign.end()},
```

Comment:
Why not move/copy constructor?

Reply:
Type of `bls_aggreate` is `std::vector<uint8_t>`, type of signature, returned from `BlsProvider` is `std::array<uint8_t, 96>`